### ISSUE TITLE
[Minor autocomplete bug?] Clang declspec properties are lower priority 

### ISSUE THREAD

**Issue by Fernthedev** [clangd/vscode-clangd#533]
`clangd` is able to recognize correctly that a property exists and is valid. However, when trying to autocomplete, it is treated as what I would call a "3rd class citizen" option that is usually resorted to VS Code's least likely useful autocomplete entry. 

My suggestion is specifically: Treat it similar to how C# treats properties prioritization.

C++:
![image](https://github.com/clangd/vscode-clangd/assets/15272073/b6e1a14f-4184-42b7-bd2c-58c379269ac3)

C#:
![image](https://github.com/clangd/vscode-clangd/assets/15272073/6263297d-b34b-446f-9433-0a60b72662a4)

**System information**
Clangd version (from the log, or `clangd --version`):
clangd extension version: 16.0.2
Operating system: Windows 10 VS Code 1.82.2 Stable

**Comment by HighCommander4** [comment]
In a minimal code example like the following:

![proposal-order](https://github.com/clangd/vscode-clangd/assets/1751085/5e162545-9c35-4db2-894f-8cd773cd54f5)

the order is fields first.

So, the issue depends on the definition of the struct in question.

Can you provide a complete code example that exhibits the undesired order?

**Comment by Fernthedev** [comment]
Sure, thanks for the quick response!
https://github.com/Fernthedev/BeatSaber-Quest-Codegen/blob/37290c93ccbe16e8a90d8573bd5d006e043a2dc1/include/UnityEngine/zzzz__GameObject_def.hpp#L78-L100

So if I'm understanding you correctly, autocomplete candidates are inversely ordered? I was looking through the source of `clangd` and noticed that properties are treated with the same score as `Variables` and `Functions` so I was very puzzled when I noticed that the closest candidate would always be last in my autocomplete.

This would make sense if it was inversely ordered 👍

**Comment by HighCommander4** [comment]
> So if I'm understanding you correctly, autocomplete candidates are inversely ordered?

I don't think so; in my minimal example, if I declare the field first, I still get the field first in the proposal list.

> https://github.com/Fernthedev/BeatSaber-Quest-Codegen/blob/37290c93ccbe16e8a90d8573bd5d006e043a2dc1/include/UnityEngine/zzzz__GameObject_def.hpp#L78-L100

What is `__declspec(property(get=get_transform))`? Is this what somehow generates the definition of a method named `get_transform`?

**Comment by Fernthedev** [comment]
Clang and MSVC both support `__declspec(property(get=getter,put=setter))`, see here https://learn.microsoft.com/en-us/cpp/cpp/property-cpp?view=msvc-170

Essentially, it's what C# offers as properties e.g 
```cs
public class Foo {
  public int X { get => getter(); set => setter(value)}
}
```

**Comment by HighCommander4** [comment]
Thanks. So to fill in some of the pieces here:

 * This is a language extension that clang supports when `-fdeclspec` is added to the compile command
 * A minimal code example demonstrating the issue would be:
     ```c++
    struct S {
       int __declspec(property(get=get_transform)) transform;
       int get_transform();
    };
    int main() {
       S s;
       int x = s.tran^
     }
     ```
    Now, regardless of the order in which the getter and the property are declared, in the proposal list the getter appears first.

I notice there are a number of other things about properties that don't work well in clangd:

 * Neither the definition of the property nor its use get a semantic highlighting token
 * Go-to-definition from the use of the property to its definition does not work

I think this is the first time this language extension has come up, and there is room for improvement in the areas listed above (and probably others I'm not thinking of) as well as completion proposal ordering.
